File: columbu/CATscala/src/main/java/com/columbu/db/levelDB/Deposit.java
package com.galileo.db.levelDB;

import com.galileo.core.gvm.vm.DataWord;
import com.galileo.core.gvm.vm.program.Storage;
import com.galileo.core.capsule.AccountCapsule;
import com.galileo.core.capsule.BlockCapsule;
import com.galileo.core.capsule.BytesCapsule;
import com.galileo.core.capsule.ContractCapsule;
import com.galileo.core.capsule.ProposalCapsule;
import com.galileo.core.capsule.TransactionCapsule;
import com.galileo.core.capsule.VotesCapsule;
import com.galileo.core.capsule.WitnessCapsule;
import com.galileo.db.levelDB.store.Manager;
import com.galileo.protos.protos.Protocol;

public interface Deposit {

  Manager getDbManager();

  AccountCapsule createAccount(byte[] address, Protocol.AccountType type);

  AccountCapsule createAccount(byte[] address, String accountName, Protocol.AccountType type);

  AccountCapsule getAccount(byte[] address);

  WitnessCapsule getWitness(byte[] address);

  VotesCapsule getVotesCapsule(byte[] address);

  ProposalCapsule getProposalCapsule(byte[] id);

  BytesCapsule getDynamic(byte[] bytesKey);

  void deleteContract(byte[] address);

  void createContract(byte[] address, ContractCapsule contractCapsule);

  ContractCapsule getContract(byte[] address);

  void saveCode(byte[] codeHash, byte[] code);

  byte[] getCode(byte[] codeHash);

  //byte[] getCodeHash(byte[] address);

  void putStorageValue(byte[] address, DataWord key, DataWord value);

  DataWord getStorageValue(byte[] address, DataWord key);

  Storage getStorage(byte[] address);

  long getBalance(byte[] address);

  long addBalance(byte[] address, long value);


  Deposit newDepositChild();

  void setParent(Deposit deposit);

  void flush();

  void commit();

  void putAccount(Key key, Value value);

  void putTransaction(Key key, Value value);

  void putBlock(Key key, Value value);

  void putWitness(Key key, Value value);

  void putCode(Key key, Value value);

  void putContract(Key key, Value value);

  void putStorage(Key key, Storage cache);

  void putVotes(Key key, Value value);

  void putProposal(Key key, Value value);

  void putDynamicProperties(Key key, Value value);

  void putAccountValue(byte[] address, AccountCapsule accountCapsule);

  void putVoteValue(byte[] address, VotesCapsule votesCapsule);

  void putProposalValue(byte[] address, ProposalCapsule proposalCapsule);

  void putDynamicPropertiesWithLatestProposalNum(long num);

  long getLatestProposalNum();

  long getWitnessAllowanceFrozenTime();

  long getMaintenanceTimeInterval();

  long getNextMaintenanceTime();

  TransactionCapsule getTransaction(byte[] trxHash);

  BlockCapsule getBlock(byte[] blockHash);

}

File: columbu/CATscala/src/main/scala/com/columbu/db/levelDB/Deposit.scala
package com.galileo.db.levelDB

import com.galileo.core.gvm.vm.DataWord
import com.galileo.core.gvm.vm.program.Storage
import com.galileo.core.capsule.AccountCapsule
import com.galileo.core.capsule.BlockCapsule
import com.galileo.core.capsule.BytesCapsule
import com.galileo.core.capsule.ContractCapsule
import com.galileo.core.capsule.ProposalCapsule
import com.galileo.core.capsule.TransactionCapsule
import com.galileo.core.capsule.VotesCapsule
import com.galileo.core.capsule.WitnessCapsule
import com.galileo.db.levelDB.store.Manager
import com.galileo.protos.protos.Protocol

trait Deposit {
  def getDbManager: Manager

  def createAccount(address: Array[Byte], `type`: Protocol.AccountType): AccountCapsule

  def createAccount(address: Array[Byte], accountName: String, `type`: Protocol.AccountType): AccountCapsule

  def getAccount(address: Array[Byte]): AccountCapsule

  def getWitness(address: Array[Byte]): WitnessCapsule

  def getVotesCapsule(address: Array[Byte]): VotesCapsule

  def getProposalCapsule(id: Array[Byte]): ProposalCapsule

  def getDynamic(bytesKey: Array[Byte]): BytesCapsule

  def deleteContract(address: Array[Byte]): Unit

  def createContract(address: Array[Byte], contractCapsule: ContractCapsule): Unit

  def getContract(address: Array[Byte]): ContractCapsule

  def saveCode(codeHash: Array[Byte], code: Array[Byte]): Unit

  def getCode(codeHash: Array[Byte]): Array[Byte]

  //byte[] getCodeHash(byte[] address);

  def putStorageValue(address: Array[Byte], key: DataWord, value: DataWord): Unit

  def getStorageValue(address: Array[Byte], key: DataWord): DataWord

  def getStorage(address: Array[Byte]): Storage

  def getBalance(address: Array[Byte]): Long

  def addBalance(address: Array[Byte], value: Long): Long


  def newDepositChild: Deposit

  def setParent(deposit: Deposit): Unit

  def flush(): Unit

  def commit(): Unit

  def putAccount(key: Key, value: Value): Unit

  def putTransaction(key: Key, value: Value): Unit

  def putBlock(key: Key, value: Value): Unit

  def putWitness(key: Key, value: Value): Unit

  def putCode(key: Key, value: Value): Unit

  def putContract(key: Key, value: Value): Unit

  def putStorage(key: Key, cache: Storage): Unit

  def putVotes(key: Key, value: Value): Unit

  def putProposal(key: Key, value: Value): Unit

  def putDynamicProperties(key: Key, value: Value): Unit

  def putAccountValue(address: Array[Byte], accountCapsule: AccountCapsule): Unit

  def putVoteValue(address: Array[Byte], votesCapsule: VotesCapsule): Unit

  def putProposalValue(address: Array[Byte], proposalCapsule: ProposalCapsule): Unit

  def putDynamicPropertiesWithLatestProposalNum(num: Long): Unit

  def getLatestProposalNum: Long

  def getWitnessAllowanceFrozenTime: Long

  def getMaintenanceTimeInterval: Long

  def getNextMaintenanceTime: Long

  def getTransaction(trxHash: Array[Byte]): TransactionCapsule

  def getBlock(blockHash: Array[Byte]): BlockCapsule

}
